### ISSUE TITLE
Avidemux cuts more accurately than losslesscut

### ISSUE THREAD

**Issue by vkopey** [mifi/lossless-cut#1786]
### I have a lot of issues to go through, so in order to make it easier for me to help you, I ask that you please try these things first

- [X] Check out all the menus and keyboard shortcuts. Maybe the feature already exists?
- [X] Try with the [newest version from GitHub](https://github.com/mifi/lossless-cut/releases/latest)
- [X] Read the [documentation](https://github.com/mifi/lossless-cut) and [FAQ, Known issues, Troubleshooting](https://github.com/mifi/lossless-cut/blob/master/issues.md) to see if the feature already exists
- [X] See if someone already suggested this under [Issues](https://github.com/mifi/lossless-cut/issues) or [Discussions](https://github.com/mifi/lossless-cut/discussions)

### Description

Has the author tried working with Avidemux? Avidemux cuts EXACTLY on key frames. But this program does not. Maybe the reason is in ffmpeg?

**Comment by mifi** [comment]
Not sure what you mean by that. Losslesscut can also cut on keyframes if you jump to keyframes when setting cutpoints

**Comment by vkopey** [comment]
Yeah, I know. But Avidemux cuts EXACTLY on the key frame and does not capture adjacent ones. Just compare the cutting quality, for example, in the MKV video AVC1+AAC.

**Comment by mifi** [comment]
LosslessCut also cuts on keyframes, because it's not possible to cut outside keyframes. Or what do you mean exactly? Do you have a concrete example of an example file with some cuts that we can compare?

**Comment by vkopey** [comment]
Sure. Here's a comparison of the results: https://drive.google.com/drive/folders/1_-s8IAHnwXejgR0XfRYKx9tU22G4FK0V?usp=sharing
Original video here: https://mega.nz/file/HahSnD5L#17tZnpZ9vd2d-6oGCxv1gcwWVJqBN04r4tiIUA8ns54

**Comment by mifi** [comment]
thanks. so the problem is that with losslesscut you get output files with an empty portion at the beginning of each file that you cannot seek to, right? i tried to change format to mp4 using losslesscut and then the result seems to be identical to avidemux. also did you use `copy` mode in avidemux?

**Comment by mifi** [comment]
related #1216

**Comment by vkopey** [comment]
Yes, you understood correctly. And I used `copy `mode in Avidemux. I will also try to cut mp4 instead of mkv and write the result here.

**Comment by Dean-Corso** [comment]
Hi guys,

so I told about that problem of key frame cuttings a longer time ago and it still exists unfortunately. First problem is displaying the right frame position in LC at all. Just load any MP4 / MKV container video which stats with any color what does change on any location or using any timer / counter anything where you can see the differences on every frame change. I made a frame counter on any video to show every single frame with a number. Now you can load the video in LC / AviDemux / VirtualDub etc you can compare it. In case of LC it starts with frame 0..
```
LC = 00:00:00.000(0)
AD = 00:00:00.000
VD = 00:00:00.000 | 0
```
...and on every first image frame I can see value 1 which is marked as 1 of course. Now when I just step forward one single frame in all tools then I should get value 2 to see what also happens in AD & VD but not in LC so this tool still does show value 1 and not value 2. LC did only add the time value of 00.040 what is the length of one frame in my example video but LC did not display the correct frame. Below 2 images...

<details>
  <summary>Frame 1</summary>

![Frame1_2023-12-03_012257](https://github.com/mifi/lossless-cut/assets/88513086/b3c1af85-aab2-4c21-84de-96e671925585)
</details>

...all starting at frame 1 time 00:00:00.000. Now when I step forward one frame we get this to see...

<details>
  <summary>Frame 2</summary>

![Frame2_2023-12-03_012257](https://github.com/mifi/lossless-cut/assets/88513086/7be4929e-dfad-4845-bccb-eff9e28c5fcf)

</details>

...so LC is still showing same frame 1 instead of frame 2 and only the time did seek forward for 0.040 (1). Why? Why does the frame not change? Now we have same problem when seeking to any key frames in the video so lets have a look.

<details>
  <summary>Key Frame Selection</summary>

![Frame3_2023-12-03_012257](https://github.com/mifi/lossless-cut/assets/88513086/f2d602ce-0327-468a-bcea-5b5fc3d71d74)
</details>

The selection seems to same on all three tools but LC does display the wrong frame (one behind). That's also confusing when user wanna do some accurate cuts by hitting the RIGHT frame and in this case the user can not hit it as you can do and see in AD & VD. 

Next problem is when you try to make a Key frame cut then it can fail when not using the right export settings. Problem I see is for example the -avoid_negative_ts option in FFmpeg. So here you can get different results out what is also confusing because the error rate is high and before you do a KF cut you should make a test cut before to know what setting to use in case of -avoid_negative_ts. Here it can fail you don't get that out what you did cut in LC. Sometimes it does not work or works wrong so this is a problem. In case of KF cuttings I still have to use AD instead of LC because it's not failing in copy mode with the videos which do fail when using ffmpeg with or without LC.

**Comment by mifi** [comment]
The problem with losslesscut showing frame 1 for the first few frames seems to be an issue with the HTML5 video player. If you do the same in Google Chrome (or even Safari), e.g. open a video file with 30fps and you seek to the second frame (using `document.getElementsByTagName('video')[0].currentTime = 0.033`), it will still show the first frame. This doesn't happen with all files and I don't know why that's happening, but essentially it means that for some videos, the frame you're seeing in the preview is not the actual frame you're at.

**Comment by mifi** [comment]
LosslessCut is first and foremost designed to make rough cuts in raw footage to cut away unused footage in order to reduce the size of files. I think even if the start/end of a video is not perfect (wrong frame number, missing a few frames, rearranged frames etc), that's ok for most people, as you can just include a little extra before/after the cut start/endpoints to make sure that you get everything. I see that it's a problem for people who cut+merge, but TBH I don't know how to fix it, because the problems don't seem to be consistent between different files, and ffmpeg and html5 video editor works differently, as mentioned above. If anyone has suggestions, I'm open to hear, but as of now there's nothing actionable for me on this.

**Comment by bryc** [comment]
Question: In LosslessCut, when you navigate to a key frame and use it as the **end time for the current segment**, is the key frame itself intended to be inclusive or exclusive?

In other words, is LosslessCut intended to include the keyframe itself as the final frame of the segment?

When cutting MP4 files (H264 + AAC), I've noticed the output often contains more frames than I intended it do. It always contains the final key frame in the output, plus some extra frames. Seems to vary, anywhere between 3 or 8 extra frames, depending on the video. It's always like this, so I've developed the technique of cutting 3-5 frames early to compensate for this and just crossing my fingers.

Looking at Avidemux, it seems that when you cut directly at a key frame, it stops on the frame right before the key frame (and does so seemingly reliably), and the final key frame itself is never included in the output.  But I do notice that sometimes a frame gets duplicated here and there in the output, which is very confusing, but I suppose it's done to pad audio frames or something like that.

It would be nice if a way to approximate the behavior of Avidemux's end marker cutting was possible, so that a user has a more reliable way to know exactly what the final frame in the output will be.

When I want more precise control over the final frame, I have to do a lot of manual checking. For example, in one random video, I had to cut **3 frames before the keyframe in LosslessCut, to match the same ending frame as Avidemux does when it cuts right on the key frame itself.**  It'd just be nice to have a more direct and reliable way of cutting the end point.

But I do understand if there are limitations to the HTML5 + FFmpeg system that might not make this possible.
